File: test/testprog.zig
const builtin = @import("builtin");
const std = @import("std");

const plthook = @import("plthook");

const lib = struct {
    extern fn strtod_cdecl(str: [*:0]const u8) f64;

    extern fn strtod_stdcall(str: [*:0]const u8) callconv(.winapi) f64;
    extern fn strtod_fastcall(str: [*:0]const u8) callconv(.{ .x86_fastcall = .{} }) f64;
    extern fn strtod_export_by_ordinal(str: [*:0]const u8) f64;
};

const OpenMode = enum(c_int) {
    open,
    open_by_handle,
    open_by_address,
};

fn showUsage() noreturn {
    std.debug.print("Usage: testprog (open | open_by_handle | open_by_address) LIB_NAME\n", .{});
    std.process.exit(1);
}

const HookedVal = struct {
    str: [30:0]u8 = .{0} ** 30,
    result: f64 = 0.0,
};

/// value captured by hook from executable to libtest.
var val_exe2lib: HookedVal = .{};
/// value captured by hook from libtest to libc.
var val_lib2libc: HookedVal = .{};

fn setResult(hv: *HookedVal, str: [*:0]const u8, result: f64) void {
    const strs = std.mem.span(str);
    @memcpy(hv.str[0..strs.len], strs);
    hv.str[strs.len] = 0;
    hv.result = result;
}

fn expectResult(str: [:0]const u8, result: f64, expected_result: f64, line: u32) !void {
    errdefer {
        std.debug.print("Error: ['{s}', {}, {}] ['{s}', {}] ['{s}', {}] at line {}\n", .{ str, result, expected_result, val_exe2lib.str, val_exe2lib.result, val_lib2libc.str, val_lib2libc.result, line });
    }

    try std.testing.expectEqual(expected_result, result);
    try std.testing.expectEqualStrings(str, std.mem.span(@as([*:0]const u8, &val_exe2lib.str)));
    try std.testing.expectEqual(expected_result, val_exe2lib.result);
    try std.testing.expectEqualStrings(str, std.mem.span(@as([*:0]const u8, &val_lib2libc.str)));
    try std.testing.expectEqual(expected_result, val_lib2libc.result);
}

fn testResult(comptime func: fn (str: [*:0]const u8) callconv(.c) f64, str: [:0]const u8, expected_result: f64, src: std.builtin.SourceLocation) !void {
    val_exe2lib = .{};
    val_lib2libc = .{};
    const result__ = func(str);
    try expectResult(str, result__, expected_result, src.line);
}

var strtod_cdecl_old_func: ?*const fn ([*:0]const u8) callconv(.c) f64 = null;

/// hook func from libtest to libc.
fn strtod_hook_func(str: [*:0]const u8) callconv(.c) f64 {
    const result = std.fmt.parseFloat(f64, std.mem.span(str)) catch 0.0;
    setResult(&val_lib2libc, str, result);
    return result;
}

/// hook func from testprog to libtest.
fn strtod_cdecl_hook_func(str: [*:0]const u8) callconv(.c) f64 {
    const result = strtod_cdecl_old_func.?(str);
    setResult(&val_exe2lib, str, result);
    return result;
}

const windows = if (builtin.os.tag == .windows) struct {
    const supports_ordinal_export = false;

    var strtod_stdcall_old_func: ?*const fn ([*:0]const u8) callconv(.winapi) f64 = null;

    /// hook func from testprog to libtest.
    fn strtod_stdcall_hook_func(str: [*:0]const u8) callconv(.winapi) f64 {
        const result = strtod_stdcall_old_func.?(str);
        setResult(&val_exe2lib, str, result);
        return result;
    }

    const x86 = if (builtin.cpu.arch == .x86) struct {
        var strtod_fastcall_old_func: ?*const fn ([*:0]const u8) callconv(.{ .x86_fastcall = .{} }) f64 = null;

        /// hook func from testprog to libtest.
        fn strtod_fastcall_hook_func(str: [*:0]const u8) callconv(.{ .x86_fastcall = .{} }) f64 {
            const result = strtod_fastcall_old_func.?(str);
            setResult(&val_exe2lib, str, result);
            return result;
        }
    };

    var strtod_export_by_ordinal_old_func: ?*const fn ([*:0]const u8) callconv(.c) f64 = if (!supports_ordinal_export) @compileError("ordinal export not supported") else null;

    /// hook func from testprog to libtest.
    fn strtod_export_by_ordinal_hook_func(str: [*:0]const u8) callconv(.c) f64 {
        if (!supports_ordinal_export) @compileError("ordinal export not supported");
        const result = strtod_export_by_ordinal_old_func.?(str);
        setResult(&val_exe2lib, str, result);
        return result;
    }
};

const EnumTestData = struct {
    name: [:0]const u8,
    enumerated: bool = false,
};

const funcs_called_by_libtest: []const EnumTestData = &.{
    if (builtin.os.tag.isDarwin()) switch (builtin.target.ptrBitWidth()) {
        32 => .{ .name = "_strtod_cust$UNIX2003" },
        64 => .{ .name = "_strtod_cust" },
        else => @compileError("Unsupported ptrBitWidth"),
    } else .{ .name = "strtod_cust" },
};

const funcs_called_by_main: []const EnumTestData = &if (builtin.os.tag == .windows and builtin.target.ptrBitWidth() == 64) .{
    .{ .name = "strtod_cdecl" },
    .{ .name = "strtod_stdcall" },
    // TODO: maybe support this
    //.{ .name = "libtest.dll:@10" },
} else if (builtin.os.tag == .windows and builtin.abi.isGnu()) .{
    .{ .name = "strtod_cdecl" },
    .{ .name = "strtod_stdcall@8" },
    .{ .name = "@strtod_fastcall@8" },
} else if (builtin.os.tag == .windows) .{
    .{ .name = "strtod_cdecl" },
    .{ .name = "_strtod_stdcall@8" },
    .{ .name = "@strtod_fastcall@8" },
    //.{ .name = "libtest.dll:@10" },
} else if (builtin.os.tag.isDarwin()) .{
    .{ .name = "_strtod_cdecl" },
} else .{
    .{ .name = "strtod_cdecl" },
};

fn test_plthook_enum(instance: *plthook.c.plthook_t, test_data: []EnumTestData) !void {
    var pos: c_uint = 0;
    var name: [*:0]const u8 = undefined;
    var addr: *?*anyopaque = undefined;

    var enumerated: usize = 0;

    while (plthook.c.plthook_enum(instance, &pos, @ptrCast(&name), @ptrCast(&addr)) == 0) {
        for (test_data) |*e| {
            if (std.mem.eql(u8, e.name, std.mem.span(name))) {
                e.enumerated = true;
                enumerated += 1;
                break;
            }
        }
    }

    if (enumerated != test_data.len) {
        for (test_data) |e| {
            if (!e.enumerated) {
                std.debug.print("{s} is not enumerated by plthook_enum.\n", .{e.name});
            }
        }
        pos = 0;
        while (plthook.c.plthook_enum(instance, &pos, @ptrCast(&name), @ptrCast(&addr)) == 0) {
            std.debug.print("   {s}\n", .{name});
        }
        return error.TestUnexpectedResult;
    }
}

fn hook_function_calls_in_executable(open_mode: OpenMode) !void {
    std.debug.print("opening executable via {}\n", .{open_mode});
    const instance = switch (open_mode) {
        .open => try plthook.openByName(null),
        .open_by_handle => blk: {
            const handle = switch (builtin.os.tag) {
                .windows => std.os.windows.kernel32.GetModuleHandleW(null).?,
                else => std.c.dlopen(null, .{ .LAZY = true }).?,
            };
            break :blk try plthook.openByHandle(handle);
        },
        .open_by_address => try plthook.openByAddress(@intFromPtr(&hook_function_calls_in_executable)),
    };
    var test_data = funcs_called_by_main[0..funcs_called_by_main.len].*;
    try test_plthook_enum(instance, &test_data);
    strtod_cdecl_old_func = try plthook.replace(instance, "strtod_cdecl", &strtod_cdecl_hook_func);
    if (builtin.os.tag == .windows) {
        windows.strtod_stdcall_old_func = try plthook.replace(instance, "strtod_stdcall", &windows.strtod_stdcall_hook_func);
        if (builtin.cpu.arch == .x86) {
            windows.x86.strtod_fastcall_old_func = try plthook.replace(instance, "strtod_fastcall", &windows.x86.strtod_fastcall_hook_func);
        }
        if (windows.supports_ordinal_export) {
            windows.strtod_export_by_ordinal_old_func = try plthook.replace(instance, "libtest.dll:@10", &windows.strtod_export_by_ordinal_hook_func);
        }
    }
    plthook.c.plthook_close(instance);
}

fn hook_function_calls_in_library(open_mode: OpenMode, filename: [:0]const u8) !void {
    std.debug.print("opening {s} via {}\n", .{ filename, open_mode });
    const instance = switch (open_mode) {
        .open => try plthook.openByName(filename),
        .open_by_handle, .open_by_address => blk: {
            const handle = switch (builtin.os.tag) {
                .windows => handle: {
                    var buf: [127:0]u16 = undefined;
                    const n = try std.unicode.utf8ToUtf16Le(&buf, filename);
                    buf[n] = 0;
                    break :handle std.os.windows.kernel32.GetModuleHandleW(buf[0..n :0]).?;
                },
                else => std.c.dlopen(filename, .{ .LAZY = true, .NOLOAD = true }).?,
            };
            break :blk switch (open_mode) {
                .open => unreachable,
                .open_by_handle => try plthook.openByHandle(handle),
                .open_by_address => blk1: {
                    const address = switch (builtin.os.tag) {
                        .windows => handle,
                        else => std.c.dlsym(handle, "strtod_cdecl").?,
                    };
                    break :blk1 try plthook.openByAddress(@intFromPtr(address));
                },
            };
        },
    };
    var test_data = funcs_called_by_libtest[0..funcs_called_by_libtest.len].*;
    try test_plthook_enum(instance, &test_data);
    _ = try plthook.replace(instance, "strtod_cust", &strtod_hook_func);
    plthook.c.plthook_close(instance);
}

pub fn main() !void {
    const gpa = std.heap.smp_allocator;
    var args = try std.process.argsWithAllocator(gpa);
    _ = args.next() orelse showUsage();
    const open_mode = std.meta.stringToEnum(OpenMode, args.next() orelse showUsage()) orelse showUsage();
    const filename = args.next() orelse showUsage();
    if (args.next()) |_| showUsage();

    const expected_result = comptime std.fmt.parseFloat(f64, "3.7") catch unreachable;

    // Resolve the function addresses by lazy binding.
    _ = lib.strtod_cdecl("3.7");
    if (builtin.os.tag == .windows) {
        _ = lib.strtod_stdcall("3.7");
        if (builtin.cpu.arch == .x86) {
            _ = lib.strtod_fastcall("3.7");
        }
        if (windows.supports_ordinal_export) {
            _ = lib.strtod_export_by_ordinal("3.7");
        }
    }

    try hook_function_calls_in_executable(open_mode);
    try hook_function_calls_in_library(open_mode, filename);

    try testResult(lib.strtod_cdecl, "3.7", expected_result, @src());
    if (builtin.os.tag == .windows) {
        try testResult(lib.strtod_stdcall, "3.7", expected_result, @src());
        if (builtin.cpu.arch == .x86) {
            try testResult(lib.strtod_fastcall, "3.7", expected_result, @src());
        }
        if (windows.supports_ordinal_export) {
            try testResult(lib.strtod_export_by_ordinal, "3.7", expected_result, @src());
        }
    }

    std.debug.print("success\n", .{});
}

File: test/testprog.c
#include <plthook.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <assert.h>
#include "libtest.h"
#ifdef _WIN32
#include <windows.h>
#else
#include <dlfcn.h>
#endif

#if defined __UCLIBC__ && !defined RTLD_NOLOAD
#define RTLD_NOLOAD 0
#endif

#define CHK_PH(func) do { \
    if (func != 0) { \
        fprintf(stderr, "%s error: %s\n", #func, plthook_error()); \
        exit(1); \
    } \
} while (0)

typedef struct {
    const char *name;
    int enumerated;
} enum_test_data_t;

enum open_mode {
    OPEN_MODE_DEFAULT,
    OPEN_MODE_BY_HANDLE,
    OPEN_MODE_BY_ADDRESS,
};

static enum_test_data_t funcs_called_by_libtest[] = {
#if defined __APPLE__ && defined __LP64__
    {"_strtod", 0},
#elif defined __APPLE__ && !defined __LP64__
    {"_strtod$UNIX2003", 0},
#else
    {"strtod", 0},
#endif
    {NULL, },
};

static enum_test_data_t funcs_called_by_main[] = {
#if defined _WIN64 || (defined __CYGWIN__ && defined __x86_64__)
    {"strtod_cdecl", 0},
    {"strtod_stdcall", 0},
    {"strtod_fastcall", 0},
#ifndef __CYGWIN__
    {"libtest.dll:@10", 0},
#endif
#elif defined _WIN32 && defined __GNUC__
    {"strtod_cdecl", 0},
    {"strtod_stdcall@8", 0},
    {"@strtod_fastcall@8", 0},
#elif defined _WIN32 && !defined __GNUC__
    {"strtod_cdecl", 0},
    {"_strtod_stdcall@8", 0},
    {"@strtod_fastcall@8", 0},
    {"libtest.dll:@10", 0},
#elif defined __APPLE__
    {"_strtod_cdecl", 0},
#else
    {"strtod_cdecl", 0},
#endif
    {NULL, },
};

typedef struct hooked_val_t hooked_val_t;

extern hooked_val_t val_exe2lib;
extern hooked_val_t val_lib2libc;

extern double (*strtod_cdecl_old_func)(const char *);
#if defined _WIN32 || defined __CYGWIN__
extern double (__stdcall *strtod_stdcall_old_func)(const char *);
extern double (__fastcall *strtod_fastcall_old_func)(const char *);
#endif
#if defined _WIN32
extern double (*strtod_export_by_ordinal_old_func)(const char *);
#endif

/* hook func from libtest to libc. */
double strtod_hook_func(const char *str);

/* hook func from testprog to libtest. */
double strtod_cdecl_hook_func(const char *str);

#if defined _WIN32 || defined __CYGWIN__
/* hook func from testprog to libtest. */
double __stdcall strtod_stdcall_hook_func(const char *str);

/* hook func from testprog to libtest. */
double __fastcall strtod_fastcall_hook_func(const char *str);
#endif

#if defined _WIN32
/* hook func from testprog to libtest. */
double strtod_export_by_ordinal_hook_func(const char *str);
#endif

static void test_plthook_enum(plthook_t *plthook, enum_test_data_t *test_data)
{
    unsigned int pos = 0;
    const char *name;
    void **addr;
    int i;

    while (plthook_enum(plthook, &pos, &name, &addr) == 0) {
        for (i = 0; test_data[i].name != NULL; i++) {
            if (strcmp(test_data[i].name, name) == 0) {
                test_data[i].enumerated = 1;
            }
        }
    }
    for (i = 0; test_data[i].name != NULL; i++) {
        if (!test_data[i].enumerated) {
            fprintf(stderr, "%s is not enumerated by plthook_enum.\n", test_data[i].name);
            pos = 0;
            while (plthook_enum(plthook, &pos, &name, &addr) == 0) {
                fprintf(stderr, "   %s\n", name);
            }
            exit(1);
        }
    }
}

void hook_function_calls_in_executable(enum open_mode open_mode)
{
    plthook_t *plthook;
    void *handle;

    fprintf(stderr, "opening executable via %d\n", open_mode);
    switch (open_mode) {
    case OPEN_MODE_DEFAULT:
        CHK_PH(plthook_open(&plthook, NULL));
        break;
    case OPEN_MODE_BY_HANDLE:
#ifdef WIN32
        handle = GetModuleHandle(NULL);
#else
        handle = dlopen(NULL, RTLD_LAZY);
#endif
        assert(handle != NULL);
        CHK_PH(plthook_open_by_handle(&plthook, handle));
        break;
    case OPEN_MODE_BY_ADDRESS:
        CHK_PH(plthook_open_by_address(&plthook, &hook_function_calls_in_executable));
        break;
    }
    test_plthook_enum(plthook, funcs_called_by_main);
    CHK_PH(plthook_replace(plthook, "strtod_cdecl", (void*)strtod_cdecl_hook_func, (void**)&strtod_cdecl_old_func));
#if defined _WIN32 || defined __CYGWIN__
    CHK_PH(plthook_replace(plthook, "strtod_stdcall", (void*)strtod_stdcall_hook_func, (void**)&strtod_stdcall_old_func));
    CHK_PH(plthook_replace(plthook, "strtod_fastcall", (void*)strtod_fastcall_hook_func, (void**)&strtod_fastcall_old_func));
#endif
#if defined _WIN32
    CHK_PH(plthook_replace(plthook, "libtest.dll:@10", (void*)strtod_export_by_ordinal_hook_func, (void**)&strtod_export_by_ordinal_old_func));
#endif
    plthook_close(plthook);
}

void hook_function_calls_in_library(enum open_mode open_mode, const char *filename)
{
    plthook_t *plthook;
    void *handle;
#ifndef WIN32
    void *address;
#endif

    fprintf(stderr, "opening %s via %d\n", filename, open_mode);
    switch (open_mode) {
    case OPEN_MODE_DEFAULT:
        CHK_PH(plthook_open(&plthook, filename));
        break;
    case OPEN_MODE_BY_HANDLE:
#ifdef WIN32
        handle = GetModuleHandle(filename);
#else
        handle = dlopen(filename, RTLD_LAZY | RTLD_NOLOAD);
#endif
        assert(handle != NULL);
        CHK_PH(plthook_open_by_handle(&plthook, handle));
        break;
    case OPEN_MODE_BY_ADDRESS:
#ifdef WIN32
        handle = GetModuleHandle(filename);
        assert(handle != NULL);
        CHK_PH(plthook_open_by_address(&plthook, handle));
#else
        handle = dlopen(filename, RTLD_LAZY | RTLD_NOLOAD);
        address = dlsym(handle, "strtod_cdecl");
        assert(address != NULL);
        CHK_PH(plthook_open_by_address(&plthook, (char*)address));
#endif
        break;
    }
    test_plthook_enum(plthook, funcs_called_by_libtest);
    CHK_PH(plthook_replace(plthook, "strtod", (void*)strtod_hook_func, NULL));
    plthook_close(plthook);
}

